content/blog/posts/2022-06-23-welcome-to-shepherd-10

```diff
@@ -0,0 +1,89 @@
+
+Shepherd 10 now supports a **callback function** that returns a DOM selector string, HTMLElement, undefined, or null as an `element` value. 
+
+```js
+const step = new Step(tour, {
+	attachTo: { element: () => { '.selector' || HTMLElement || '[data-test-...]' || undefined || null },
+	on: 'top'}
+}) 
+```
+
+We now resolve the `attachTo.element` in the `before-show` phase which by its name resolves the attachTo specifications just once before the Shepherd Step is shown on the DOM. By also **memoizing** the attachTo resolution, the Shepherd 10 delivers a more performant experience to users as they navigate each step in a tour.
+
+## Memo-what?
+
+If you haven't heard of memoization or memoize before, not to worry it is simply a way of storing information which in most cases saves time.
+
+> In programming, **memoization is an optimization technique** that makes applications more efficient and hence faster. It does this by storing computation results in cache, and retrieving that same information from the cache the next time it's needed instead of computing it again. --[Germán Cocca](https://www.freecodecamp.org/news/author/gercocca/)
+
+<iframe src="https://gifer.com/embed/EVRC" width=480 height=276.480 frameBorder="0" allowFullScreen></iframe>
+
+It's not surprising that the words `memorize` and `memoize` share the same latin root. Instead of constantly wondering what is in the fridge, writing down your list and carrying it with you to the grocery store is a much cleaner active recall. For those familair with complexity, memoization is often described as a trade of space complexity for time complexity. We take up a bit more space in the cache with the intention of faster load time.
+
+## In Action
+
+For those familiar with Shepherd, very little has changed about how you would define the attachTo element. We still require some element to place the step (or null or undefined), but using the callback opens up our options for placing the step.
+
+A simple example, we have a button that exapands a custom dropdown menu. However because menu options are dynamic, they can vary from user to user or login state.
+```html
+<button class="menu">
+	<a class="apples">apples</a>
+	<a class="pears">pears</a>
+	<!-- etc.. -->
+</button>
+```
+In the tour, let's say we want a Step to eventually attach to `oranges`. Let's configure the callback to select it by its class name.
+```html
+<!-- Sometimes will be on the DOM, not always -->
+<a class="oranges">oranges</a>
+```
+
+```js
+const step = new Step(tour, {
+	attachTo: { element: () => { '.oranges' },
+	on: 'bottom'}
+}) 
+``` 
+
+And that's it! As long as the callback return value is one of the accepted data types our Step will attach as expected even to dynamic elements. In previous iterations of Shepherd this would not work because the oranges element wouldn't exist at the time of attachTo resolution. Let's take a look at another way we can select  `oranges`.
```

Comment:
some syntax/formatting:

As long as the callback's return value is one of the accepted data types, our `Step` will attach as expected -- even to dynamic elements. In previous iterations of Shepherd, this would not have worked because the `oranges` element would not yet exist at the time of `attachTo` resolution. Let's take a look at another way we can select `oranges`: